### PR TITLE
feat(sdk): add query DSL to express conditions and options

### PR DESCRIPTION
**Pull request by oyyblin**

### CODE REVIEW

**Review by Stumble — COMMENTED** [review]
Let's update both tutorials?

**Review by Stumble — CHANGES_REQUESTED** [review]
some completeness and validity checks are missing.

File: packages/sdk/src/roles/user.ts [hunk 1]
```diff
@@ -203,4 +205,73 @@ export class User {
...
+              BigInt((expr.value as query.Range).from),
+              BigInt((expr.value as query.Range).to)
+            )
+          );
+          break;
+        }
+        case ClaimTypeEnum.Property: {
+          statements.push(new statement.PropStatement(claim.type as PropType, expr.value as query.Set));
+          break;
+        }
```
[the remark this answers is not in the document]
Comment (oyyblin) [reply packages/sdk/src/roles/user.ts]: updated

File: packages/sdk/src/roles/user.ts [hunk 2]
```diff
@@ -203,4 +205,73 @@ export class User {
...
+      throw new Error("No identity slice found");
+    }
+
+    const queryObj = query.parse(queryStr);
+    const statements: statement.Statement[] = [];
+    for (const expr of queryObj.conditions) {
+      const claim = cred.body.tp.claims.find(c => c.name === expr.identifier);
+      if (!claim) {
+        throw new Error("Claim not found in the credential body");
+      }
```
[the remark this answers is not in the document]
Comment (oyyblin) [reply packages/sdk/src/roles/user.ts]: added claims / statements count check

File: packages/sdk/src/roles/user.ts [hunk 3]
```diff
@@ -203,4 +203,89 @@ export class User {
...
+        }
+        default:
+          throw new Error("Unsupported claim type");
+      }
+    }
+
+    // Ensure all claims have corresponding statements.
+    if (cred.body.tp.claims.length !== statements.length) {
+      throw new Error("Claims and statements mismatch");
+    }
```
Comment (Stumble) [inline packages/sdk/src/roles/user.ts:276]: not sound? Must check 1-to-1 mapping. One can have a query looks like this:

"A in [1,2];A in [1,2];A in [1,2];A in [1,2];A in [1,2];A in [1,2];"

File: packages/sdk/src/roles/user.ts [hunk 4]
```diff
@@ -203,4 +203,89 @@ export class User {
...
+  ): Promise<WholeProof> {
+    const userSecrets = this.getIdentitySliceByIdc(identityCommitment);
+    if (!userSecrets) {
+      throw new Error("No identity slice found");
+    }
+
+    const queryObj = query.parse(queryStr);
+    if (!queryObj.ok) {
+      throw queryObj.error;
+    }
```
Comment (Stumble) [inline packages/sdk/src/roles/user.ts:228]: use unwrap? unwrap(query.parse(queryStr))

Comment (oyyblin) [reply packages/sdk/src/roles/user.ts:228]: updated. nice catch

File: packages/sdk/src/roles/user.ts (hunk 3 above)
Comment (oyyblin) [reply packages/sdk/src/roles/user.ts:276]: this is checked in query.ts. But for clarity, updated validations, and updated error messages and tests

**Review by Stumble — APPROVED** [review]

**Pull request merged**